kayak_render_macros/src/use_effect.rs

```diff
@@ -0,0 +1,67 @@
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use quote::{format_ident, quote};
+use syn::{bracketed, Token};
+use syn::parse::{Parse, ParseStream};
+use syn::punctuated::{Iter, Punctuated};
+
+pub(crate) struct UseEffect {
+    pub closure: syn::ExprClosure,
+    pub dependencies: Punctuated<Ident, Token![,]>,
+}
+
+impl Parse for UseEffect {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let raw_deps;
+        let closure = input.parse()?;
+        let _: Token![,] = input.parse()?;
+        let _ = bracketed!(raw_deps in input);
+        let dependencies = raw_deps.parse_terminated(Ident::parse)?;
+
+        Ok(Self {
+            closure,
+            dependencies,
+        })
+    }
+}
+
+impl UseEffect {
+    fn get_deps(&self) -> Iter<Ident> {
+        self.dependencies.iter()
+    }
+
+    fn get_clone_dep_idents(&self) -> impl Iterator<Item=Ident> + '_ {
+        self.get_deps().map(|dep| format_ident!("{}_dependency_clone", dep))
+    }
+
+    fn create_clone_deps(&self) -> proc_macro2::TokenStream {
+        let deps = self.get_deps();
+        let cloned_deps = self.get_clone_dep_idents();
+        quote! {
+            #(let #cloned_deps = #deps.clone());*
+        }
+    }
+
+    fn create_dep_array(&self) -> proc_macro2::TokenStream {
+        let cloned_deps = self.get_clone_dep_idents();
+        quote! {
+            &[#(&#cloned_deps),*]
+        }
+    }
+
+    /// Build the output token stream, creating the actual use_effect code
+    pub fn build(self) -> TokenStream {
+        let dep_array = self.create_dep_array();
+        let cloned_deps = self.create_clone_deps();
+        let closure = self.closure;
+        let result = quote! {{
+            use kayak_core::{Bound, MutableBound};
+            #cloned_deps;
```

Comment:
Maybe I should also clone the dependencies (without the custom naming) so users don't have to do it themselves when using a state internally in the closure. Then again that might be unintuitive and would probably contain unnecessary clones? But it could also streamline the usage? Hm 🤔

Reply:
I think this looks fine as is.